x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Greatest.java

```diff
@@ -111,6 +111,8 @@ public Object fold() {
 
     @Override
     public ExpressionEvaluator.Factory toEvaluator(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
+        // force datatype initialization
+        var dataType = dataType();
```

Comment:
Wondering why this wasn't needed so far. 
Nit: maybe just calling the function ignoring the return and not shadowing the instance variable might be cleaner.

Reply:
This was a bug - the node was mutated but the new instance did not have its dataType()) called.
For example, the node was serialized, deserialized than the execution kicked in - since nobody called dataType(), the value was not initialized.